File: linked-list/linkedList.cpp
#include <iostream>

using namespace std;


/*

A node in a singly linked list has the following properties: data and next. data
is the value for the linked list node, and next is a pointer to another instance of 
ListNode

*/


template <class T>
class Node
{
public:
    T data;
    Node<T>* next;
};

/*

A linked list, in its simplest form, is a collection of nodes that together form a
linear ordering.
*/

template <class T>
class LinkedList
{

private:
    /*
        This can be referred as Root or Head of the Nodes
    */
    Node<T>* root;

public:

    LinkedList();
    ~LinkedList();

    bool isEmpty();
    void addFront(const T& elem); // //add elem to the front of the linked list
    void addBack(const T& elem); // add elem to the back of the Linked list;

    T& back(); //returns the end or back of the linked list node;
    T& front(); //returns the front of the linked list node;

    int removeBack(); // removes node from the back or at the end of the list;
    int removeFront(); // removes node from the front
    void print();
};


template <class T> LinkedList<T>::LinkedList()
{
    cout<<"Initialized"<<endl;
    root = NULL;
}

template <class T> LinkedList<T>::~LinkedList()
{
    while(!isEmpty())
    {
        removeFront();
    }
}

template <class T> T& LinkedList<T>::front()
{
    return root->data;
}

template <class T> T& LinkedList<T>::back()
{
    Node<T>* tmp = root;
    while(!tmp->next==NULL)
    {
        tmp = tmp->next;
    }

    return tmp->data;
}

template <class T> bool LinkedList<T>::isEmpty()
{
    return root==NULL;
}

template <class T> void LinkedList<T>::addFront(const T& elem)
{
    Node<T>* newNode = new Node<T>;
    newNode->data = elem;
    newNode->next = root;
    root = newNode;
}
template <class T> void LinkedList<T>::addBack(const T& elem)
{
    Node<T>* newNode = new Node<T>;
    newNode->data = elem;
    newNode->next = NULL;

    Node<T>* tmp = root;
    while(!tmp->next==NULL)
    {
        tmp = tmp->next;
    }

    tmp->next = newNode;
}




/*
    Delete the front element and assign the rest to root
*/

template <class T> int LinkedList<T>::removeFront()
{
    if(isEmpty())
        return 0;
    Node<T>* old = root;
    root = old->next;
    delete old;
}

template <class T> int LinkedList<T>::removeBack()
{
    if(isEmpty())
        return 0;
    Node<T>* old = root;
    Node<T>* prev = NULL; //Used to set last element next to NULL;
    while(!old->next==NULL){
        prev = old;
        old= old->next;
    }
     prev->next = NULL;

    cout<<"Removed : " <<old->data<<endl;
    delete old;
}

/*
    print Linked list element while traversing the nodes
*/
template <class T> void LinkedList<T>::print()
{
    Node<T>* tmp = root;
    while(!tmp==NULL)
    {
        cout<<tmp->data<<"\t";
        tmp = tmp->next;
    }
    cout<<endl;
}

/*
int main()
{
    LinkedList<int> ll;
    ll.addFront(1);
    ll.addFront(2);
    ll.addFront(3);
    ll.addBack(0);
    ll.print();  //3 2 1 0
    ll.removeBack();
    ll.removeFront();
    ll.print(); // 2 1

    ll.addFront(3);
    ll.addBack(0);
    ll.print(); // 3 2 1

}

*/

File: linked-list/linkedList.js
/*

A node in a singly linked list has the following properties: data and next. data
is the value for the linked list node, and next is a pointer to another instance of 
ListNode

*/

function ListNode(data) {
  this.next = null;
  this.data = data;
}

/*

A linked list, in its simplest form, is a collection of nodes that together form a
linear ordering.
*/

class LinkedList {
  constructor() {
    this._root = null;
    this.count = 0;
  }

  /* insert Node at the end */
  insert(data) {
    let temp = this._root;
    if (this._root == null) {
      this._root = new ListNode(data);
      this.count++;
      return;
    }
    while (true) {
      if (temp.next == null) {
        temp.next = new ListNode(data);
        this.count++;
        break;
      }
      temp = temp.next;
    }
  }

  /* insert Node at the front */

  insertFront(data) {
    let newNode = new ListNode(data); //data->null;
    newNode.next = this._root; // data->root;
    this._root = newNode;
    this.count++;
  }

  /* Delete a node with a given value and return a bool response */
  delete(value) {
    let temp = this._root;
    if (temp == null) console.log("Empty");
    let prevNode = temp; // a node previous to the node to be deleted; eg) 4 in 4->5->6  while 5 to be deleted

    /* Delete first elemnt check */
    if (this._root.data == value) {
      console.log("A node deleted successfully!");
      this._root = this._root.next;
      this.count--;
      return true;
    }

    while (temp != null) {
      if (temp.data == value) {
        console.log("A node deleted successfully!");
        prevNode.next = temp.next;

        delete temp.data;
        temp.next = null;

        this.count--;

        return true;
      }
      prevNode = temp;
      temp = temp.next;
    }
    console.log("A node is not found!");
    return false;
  }

  /* size of the LinkedList */
  size() {
    return this.count;
  }

  /* Traverse the linked list while printing data */
  print() {
    let temp = this._root;
    if (temp == null) console.log("Empty");
    while (temp != null) {
      console.log(temp.data);
      temp = temp.next;
    }
  }
}

let ll = new LinkedList();
ll.insert(1);
ll.insert(2);
ll.insert(3);
ll.insert(4);
ll.insert(5);
ll.insert(6);
ll.insert(7);
ll.insert(8);

ll.insertFront(0);
ll.delete(0);
ll.delete(4);
ll.print();
